harness/determined/common/storage/shared.py

```diff
@@ -3,134 +3,116 @@
-
-        for srcentry in entries:
-            if srcentry.name in ignored_names:
-                continue
-            srcname = os.path.join(src, srcentry.name)
-            dstname = os.path.join(dst, srcentry.name)
-            srcobj = srcentry if use_srcentry else srcname
-            try:
-                is_symlink = srcentry.is_symlink()
-                if is_symlink and os.name == "nt":
-                    # Special check for directory junctions, which appear as
-                    # symlinks but we want to recurse.
-                    lstat = srcentry.stat(follow_symlinks=False)
-                    if lstat.st_reparse_tag == stat.IO_REPARSE_TAG_MOUNT_POINT:
-                        is_symlink = False
-                if is_symlink:
-                    linkto = os.readlink(srcname)
-                    if symlinks:
-                        # We can't just leave it to `copy_function` because legacy
-                        # code with a custom `copy_function` may rely on copytree
-                        # doing the right thing.
-                        os.symlink(linkto, dstname)
-                        shutil.copystat(srcobj, dstname, follow_symlinks=not symlinks)
-                    else:
-                        # ignore dangling symlink if the flag is on
-                        if not os.path.exists(linkto) and ignore_dangling_symlinks:
-                            continue
-                        # otherwise let the copy occur. copy2 will raise an error
-                        if srcentry.is_dir():
-                            copytree(
-                                srcobj,
-                                dstname,
-                                symlinks,
-                                ignore,
-                                copy_function,
-                                dirs_exist_ok=dirs_exist_ok,
-                            )
-                        else:
-                            copy_function(srcobj, dstname)
-                elif srcentry.is_dir():
+# Based on shutil.copytree and shutil._copytree (for Python 3.8). Compared to the original
+# implementation this code delays creating new directories during traversal, such that dir
+# is created only when (1) selector(dir)==True or (2) selector(dir/subdir/path)==True.
+# The code is simplified to rely on default values for:
+# symlinks=False,
+# ignore=None,
+# copy_function=shutil.copy2,
+# ignore_dangling_symlinks=False.
+
+
+def _copytree(
+    entries: List,
+    src: str,
+    dst: str,
+    selector: Optional[Callable[[str], bool]],
+    src_root: str,
+    dirs_exist_ok: bool = False,
+) -> str:
+    errors = []
+    have_copied = False
+    for srcobj in entries:
+        srcname = os.path.join(src, srcobj.name)
+        dstname = os.path.join(dst, srcobj.name)
+        src_relpath = os.path.relpath(srcname, src_root)
+        try:
+            is_symlink = srcobj.is_symlink()
+            if is_symlink:
+                if srcobj.is_dir():
```

Comment:
something is wrong with this... I don't think it's possible for `srcobj.is_symlink()` and `srcobj.is_dir()` can both return `True`.

Reply:
It's possible :) I added a test in test_shared_fs.py showing how to get there. Also, since there is absolutely no difference between behavior for symlinks and non-symlink, copytree is simplified even further.

Reply:
Well, TIL.